[CodeGen] Use LocationSize for MMO getSize

davemgreen opened the pull request:
This is part of #70452 that changes the type used for the external interface of MMO to LocationSize as opposed to uint64_t. This means the constructors take LocationSize, and convert ~UINT64_C(0) to LocationSize::beforeOrAfter(). The getSize methods return a LocationSize.

This allows us to be more precise with unknown sizes, not accidentally treating them as unsigned values, and in the future should allow us to add proper scalable vector support but none of that is included in this patch. It should mostly be an NFC.

Global ISel is still expected to use the underlying LLT as it needs, and are not expected to see unknown sizes for generic operations. Most of the changes are hopefully fairly mechanical, adding a lot of getValue() calls and protecting them with hasValue() where needed.

arsenm commented on llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h:
```diff
@@ -647,15 +647,15 @@ bool GIMatchTableExecutor::executeMatchTable(
 
       unsigned Size = MRI.getType(MO.getReg()).getSizeInBits();
```
should Size not be unsigned here?

arsenm commented on llvm/include/llvm/CodeGen/MachineFunction.h:
```diff
@@ -1046,9 +1055,16 @@ class LLVM_EXTERNAL_VISIBILITY MachineFunction {
     return getMachineMemOperand(
-        MMO, Offset, Size == ~UINT64_C(0) ? LLT() : LLT::scalar(8 * Size));
+        MMO, Offset,
+        !Size.hasValue() || Size.isScalable()
```
Can this just assert Size.hasValue()? I would assume if you have the unknown case you would have been using the LLT constructor in the first place

davemgreen replied:
I think quite a lot of things will construct an MMO with an unknown LocationSize, either directly or from the size in another MMO that was already unknown.

~UINT64_C(0) and MemoryLocation::UnknownSize was used a lot in the past, they have all been changed to LocationSize::beforeOrAfterPointer() as the unknown Size.

davemgreen committed:
[CodeGen] Use LocationSize for MMO getSize

This is part of #70452 that changes the type used for the external interface of
MMO to LocationSize as opposed to uint64_t. This means the constructors take
LocationSize, and convert ~UINT64_C(0) to LocationSize::beforeOrAfter(). The
getSize methods return a LocationSize.

This allows us to be more precise with unknown sizes, not accidentally treating
them as unsigned values, and in the future should allow us to add proper
scalable vector support but none of that is included in this patch. It should
mostly be an NFC.

Global ISel is still expected to use the underlying LLT as it needs, and are
not expected to see unknown sizes for generic operations. Most of the changes
are hopefully fairly mechanical.

davemgreen commented:
Thanks.

I've ran some testing, which all checked out, but this has a chance of hitting asserts in cases that were previously just converted to int. Let me know if this runs into any problems.
